### ISSUE TITLE
New hostname config UX

### ISSUE THREAD

**Issue by vmuzikar** [keycloak/keycloak#24252]
# Motivation

Keycloak has offered [Hostname SPI](https://github.com/keycloak/keycloak/blob/4bcc5df8585257c2b6282d421f142e27b3a38443/server-spi/src/main/java/org/keycloak/urls/HostnameProvider.java) for many years now. The SPI is critical for secure Keycloak deployments as it allows the server to determine URLs (incl. hostname, port, scheme, context path) based on various factors like incoming request or hardcoded parts of the URL reflecting the requested URL type (frontend, backend, admin). This is essential from security perspective as request headers can be forget by an attacker.

With the switch to the Quarkus distribution, a new [implementation](https://github.com/keycloak/keycloak/blob/4bcc5df8585257c2b6282d421f142e27b3a38443/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/hostname/DefaultHostnameProvider.java) of this SPI was introduced to align with the new configuration experience. The implementation attempted to provide wide variety of options with some smart and secure-by-default defaults that were based on other settings (like the `proxy` option). However, due to its complexity it quickly became very confusing and frustrating for users to configure and also suffered from many [bugs](https://github.com/keycloak/keycloak/issues/14666). It became challeging even for Keycloak developers to understand the intended behaviour hence becoming a maintanence burden.

Fixing the current implementation would not be possible without introducing significant breaking changes, hence breaking it even more for users that finally managed to configure it properly for their use case.

To resolve this, we would like to explore the possibility of introducing a brand new implementation incl. new simplified config options. This would be independent of the current implementation of the [Hostname SPI](https://github.com/keycloak/keycloak/blob/4bcc5df8585257c2b6282d421f142e27b3a38443/server-spi/src/main/java/org/keycloak/urls/HostnameProvider.java) and it'd offer the possibility of deprecation period before removing the old implementation.

# Background

The Hostname SPI recognizes the following URL types:
* **Frontend.** Used for in-browser authentication flows, account console, etc.
* **Backend.** Used for backend communication between a client (an application secured by Keycloak) and Keycloak itself. This might be over a local network,  avoiding a reverse proxy.
* **Admin Console.** A URL used for Admin Console. It is used for redirect URLs, loading resources (CSS, JS), etc.

Sources for resolving the URL:
* Dynamic from an incoming request
  * `Host` header, scheme, server port, context path
  * Proxy-set headers: `Forwarded` and `X-Forwarded-*` headers
* Hardcoded
    * Server-wide config (currently e.g `--hostname`, `--proxy`, etc.)
    * Realm configuration for frontend URL

# New config options proposal

The aim is to provide clarity to how the URLs are resolved (what source is used), remove some opaque settings (like strict HTTPS scheme or ambiguous proxy setting) and simplify the implementation to improve maintainability making it less error prone. The intention is to strongly suggest to the user to hardcode full URLs whenever possible. This will improve security as well as make the configuration more straightforward. And whenever hardcoding is not a feasible option, making it clear that dynamic sources are used.

* `--hostname hostname|URL`
  * Accepts either a hostname, or full URL incl. a port number, scheme and context root.
  * Affected URL types:
    * Frontend, unless overriden on Realm level.
    * Backend IF `--hostname-backchannel-dynamic` is set to `false` (default).
    * Admin Console IF `--hostname-admin` is not specified.
  * Possible values:
    * Full URL: the hardcoded URL value is used, no dynamic resolution whatsoever. Example: `--hostname https://my.keycloak.org:8443/auth`.
    * Hostname only: hostname is hardcoded and the rest (scheme, port, etc.) is resolved dynamically from the request. Example: `--hostname my.keycloak.org`.
* `--hostname-admin URL`
  * Affected URL types:
    * Admin Console.
  * Possible values:
    * Full URL: the hardcoded URL value is used, no dynamic resolution whatsoever. Example: `--hostname-admin https://admin.my.keycloak.org:8443/auth`.
* `--hostname-backchannel-dynamic true|false`
  * Affected URL types:
    * Backend.
  * Possible values:
    * `false` (default): backend is resolved as if it was frontend.
    * `true`: backend is resolved dynamically based on the incoming headers incl. proxy headers if `--proxy-headers` is configured.
* `--hostname-strict true|false`
  * Configures whether Keycloak requires explicitly set hostname.
  * Affected URL types:
    * Frontend when `--hostname` is undefined.
    * Backend when `--hostname-backchannel-dynamic false` (and therefore resolving as frontend) AND `--hostname` is undefined.
    * Admin Console when `--hostname-admin` (and therefore resolving as frontend) AND `--hostname` are both undefined.
  * Possible values:
    * `true` (default): Keycloak requires a hardcoded hostname via the `--hostname` option.
    * `false`: Keycloak does not need a hardcoded hostname, hence it is possible to omit the `--hostname` option. Hostname is then resolved dynamically either through the `host` header or `Forwarded`/`X-Forwarded-*` headers.
* `--proxy-headers forwarded|xforwarded`
  * Configures Keycloak to accept headers set by a reverse proxy.
  * Affected URL types:
    * Frontend (excl. hostname) when `--hostname hostname` (NOT when `--hostname URL`).
    * Backend when `--hostname-backchannel-dynamic true` OR when `--hostname hostname` (excl. hostname).
    * Admin Console when `--hostname hostname` AND `--hostname-admin` are undefined.
  * Also affects the client IP address as it's seen by Keycloak.
  * Possible values:
    * `forwarded`: Keycloak accepts the [`Forwarded` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded) to resolve hostname, client IP and protocol
    * `xforwarded`: Keycloak accepts the `X-Forwarded-*` headers, namely:
      * `X-Forwarded-Proto`
      * `X-Forwarded-Host`
      * `X-Forwarded-Port`
      * `X-Forwarded-Ssl`
      * `X-Forwarded-Prefix`
  * See also: https://quarkus.io/guides/http-reference#reverse-proxy

## Validations
* `--hostname URL` and `--hostname-admin URL` are verified that full URL is used, incl. scheme and hostname. Port is validated only if present, otherwise default port for given protocol is assumed (80 or 443).
* In production profile (`kc.sh|bat start`), either `--hostname` or `--hostname-strict false` MUST be explicitly configured.
  * Does not apply for dev profile (`kc.sh|bat start-dev`) where  `--hostname-strict false` is the default.
* If `--hostname` is not configured:
  * `--hostname-backchannel-dynamic` MUST be set to `false`.
  * `--hostname-strict` MUST be set to `false`.
* If `--hostname` is configured, `--hostname-strict` MUST be set to `true`.
* If `--hostname-admin` is configured, `--hostname` MUST be set to a URL (not just `hostname`).
* If `--hostname-backchannel-dynamic` is set to `true`, `--hostname` MUST be set to a URL (not just `hostname`).

### I want to start Keycloak on my hostname, I don't use any proxy
```
kc.sh start --hostname my.keycloak.org
```
* I can access Keycloak via `https://my.keycloak.org:8443` (no HTTP since I haven't explicitly enabled the HTTP listener). Using the default HTTPS listener port 8443 as I haven't configured a full URL in `--hostname`.
* All endpoints use `my.keycloak.org`.

### I have a TLS-terminating proxy that does not set proper headers
```
kc.sh start --hostname https://my.keycloak.org --http-enabled true
```
* I can access Keycloak via `https://my.keycloak.org`, proxy then accesses Keycloak via HTTP.
* All endpoints use `https://my.keycloak.org`.

### I have a proxy and I trust it sets `Forwarded` header correctly, I don't want to hardcode a hostname
```
kc.sh start --hostname-strict false --proxy-headers forwarded
```
* I can access Keycloak via whatever my proxy sets in `Forwarded`.
* All endpoints use the header and dynamic URL resolution.

### I want to expose Admin Console on a different host
```
kc.sh start --hostname https://my.keycloak.org --hostname-admin https://admin.my.keycloak.org:8443
```
* I can access Keycloak via `https://my.keycloak.org`.
* I can access Keycloak Admin Console via `https://admin.my.keycloak.org:8443`.
* Backend uses `https://my.keycloak.org`.

### I have a proxy but I want my apps to access Keycloak via my local network
```
kc.sh start --hostname https://my.keycloak.org --hostname-backchannel-dynamic true
```
* I can access Keycloak via `https://my.keycloak.org`.
* Backend uses dynamic URL resolution.

# Deprecation proposal for hostname options

* Keycloak 25
  * Introduce the new config options with new hostname implementation. It will be enabled by default with the possibility to disable it via `--features hostname:v1` or `--features-disabled hostname:v2`.
    * When enabled, old options (as listed below) and old Hostname SPI implementation are deactivated.
  * Mark old options as deprecated.
* Keycloak 27
  * Remove `hostname2` feature making the new implementation and options the only one available.
  * Remove old options and implementation.

## Removed old options

* `--hostname-url`
* `--hostname-admin-url`
* `--hostname-strict-https`
* `--hostname-strict-backchannel`
* `--hostname-path`
* `--hostname-port`

# Deprecation proposal for proxy option

* Keycloak 24
  * Introduce the new `--proxy-headers` option.
  * Mark the `--proxy` option as deprecated.
* Keycloak 27
  * Remove the `--proxy` option.

# Issues

* [x] https://github.com/keycloak/keycloak/issues/13113
* [x] https://github.com/keycloak/keycloak/issues/23431
* [x] https://github.com/keycloak/keycloak/issues/24668
* [x] #25336
* [x] https://github.com/keycloak/keycloak/issues/26084
* [x] https://github.com/keycloak/keycloak/issues/27729
* [x] https://github.com/keycloak/keycloak/issues/27728
* [x] https://github.com/keycloak/keycloak/issues/27730

Follow-up issue:
* [ ] https://github.com/keycloak/keycloak/issues/27731

**Previously created issues concerning the previous Hostname SPI implementation that will get resolved in the new implementation:**

* https://github.com/keycloak/keycloak/issues/23411
* https://github.com/keycloak/keycloak/issues/14666
* https://github.com/keycloak/keycloak/issues/22116
* https://github.com/keycloak/keycloak/issues/16301

**Comment by stianst** [comment]
As this has support for setting what headers are considered in the host config there's less use of `proxy`. However, there's the X-Forwarded-For header which is not really a hostname thing, so how would you configure Keycloak to use the X-Forwarded-For header? It's also a little bit strange to configure Forwarded header has part of `host` config as it really does two things (URL + source IP).

**Comment by stianst** [comment]
Validations: `are verified that full URL is used, incl. scheme, hostname, port.` port can be omitted if the default port is used, so we need to make sure the validation permits `https://mykeycloak`

**Comment by stianst** [comment]
Before we go ahead I would actually like to see the way this would be documented first. Basically write the new hostname config guide first ;)

**Comment by vmuzikar** [comment]
@stianst Thanks for the feedback!

---

> * In prod mode only https is enabled by default (secure by default)

Yes, [`http-enabled`](https://github.com/keycloak/keycloak/blob/54a081832a3bfd81470b3382548e678621506f0c/quarkus/config-api/src/main/java/org/keycloak/config/HttpOptions.java#L11) is set to `false` by default.

---

> As this has support for setting what headers are considered in the host config there's less use of `proxy`.

Sorry, I forgot to list `--proxy` option among the deprecated options. Fixed that now.

---

> I'd suggest a short name for `--host-dynamic-headers`, maybe just `--host-headers`?

> It's also a little bit strange to configure Forwarded header has part of `host` config as it really does two things (URL + source IP).

This is about naming the `--host-dynamic-headers` option. Maybe rather `--http-accept-headers`? We want to also cover the use case when user explicitly enables `Host` headers resolution without specifying `--host` option, so it's not just proxy related.

---

> Validations: `are verified that full URL is used, incl. scheme, hostname, port.` port can be omitted if the default port is used, so we need to make sure the validation permits `https://mykeycloak` (and no I don't want to have tho include default ports when I'm setting this, as shown in the examples ;))

I'd like to go with super explicit way in configuration to avoid ambiguousness which was one of the issues with the previous hostname impl. Of course in final resolved URLs the default ports would not be present (as shown in examples). :)

**Comment by stianst** [comment]
Dynamic based on proxy headers or host header for that matter only works if admin + backend = frontend. Otherwise there are cases where both backend and admin need to resolve frontend that won't work properly.

**Comment by vmuzikar** [comment]
Based on an offline discussion with @stianst I made the following updates:
* Renamed options from `--host-*` to `--hostname-*` as most of them remain.
* Replace `--host-dynamic-headers` with `--proxy-headers` and `--hostname-strict`. The split was done because `[X-]Forwarded-*` headers do not affect only URLs resolution (Hostname SPI) but also are used for determining Client IP.
  * Updated the validations to reflect this change.
* Updated deprecation plan for the `--proxy` option as it's independent from the Hostname implementation.

**Comment by stianst** [comment]
Examples should not use default ports, so either remove default ports, or use a non-default port like 8443.

`--hostname-backchannel-dynamic` states it is dynamically resolved based on `--hostname-dynamic-headers`, but that doesn't exist anymore, so guess replace with `--proxy-headers`?

Should make it clear that default ports will not be included in generated URLs by the hostname provider.

If `--hostname` is not configured why does it say `hostname-backchannel-dynamic` must be `false`? Would have thought it doesn't matter when frontend is dynamic, as backend would be dynamic regardless.

Examples `I want to start Keycloak on my hostname, I don't use any proxy` is missleading as it would actually be `https://my.keycloak.org:8443`, unless `https-port` is specified.

**Comment by stianst** [comment]
I wonder if we should extend features stuff to support versioning of providers. It's a bit of an oddity that you enable the `v2` with `--features=hostname2`, and after it becomes the default that you use `v1` with `--features-disabled=hostname2`. Perhaps we could do something like:

`--features=hostname:v2`
`--features=hostname:v1`

For a given feature we'd have a list of versions, where a given version can be default, preview, experimental, or deprecated.

**Comment by vmuzikar** [comment]
@stianst Thanks for the input.

> Examples should not use default ports, so either remove default ports, or use a non-default port like 8443.

Fixed.

> `--hostname-backchannel-dynamic` states it is dynamically resolved based on `--hostname-dynamic-headers`, but that doesn't exist anymore, so guess replace with `--proxy-headers`?

Good catch! Fixed.

> Should make it clear that default ports will not be included in generated URLs by the hostname provider.

This is hinted in the Validations.
```
otherwise default port for given protocol is assumed (80 or 443)
```
Do you want to make it more explicit?

> If `--hostname` is not configured why does it say `hostname-backchannel-dynamic` must be `false`? Would have thought it doesn't matter when frontend is dynamic, as backend would be dynamic regardless.

Because:
```
If `--hostname-backchannel-dynamic` is set to `true`, `--hostname` MUST be set to a URL (not just `hostname`).
```
I'd like to avoid any ambiguity and don't leave any room to potentially inconsistent config.

> Examples `I want to start Keycloak on my hostname, I don't use any proxy` is missleading as it would actually be `https://my.keycloak.org:8443`, unless `https-port` is specified.

Good catch. Fixed as well.

> I wonder if we should extend features stuff to support versioning.

Created https://github.com/keycloak/keycloak/issues/24668.

**Comment by sschu** [comment]
I haven't given this a thorough review, yet. I would just like to put one more requirement here. We have customers that after some time become unhappy with the hostname/domain they picked and want to migrate it to a new name. It would be great if it was possible to pick two different names in parallel to facilitate that migration. This way customers can just switch their clients from the old to the new name one after another and not in a big bang, which is hardly possible in practice.

**Comment by vmuzikar** [comment]
> It would be great if it was possible to pick two different names in parallel to facilitate that migration.

This is, unfortunately, not possible from security perspective. If proxy headers are not used (and of course neither the `host` header which should be used only in dev environment), we have to hardcode the hostname in links. Otherwise we risk that an attacker could e.g. make KC to send a password reset email with links that point to a malicious hostname that is under attacker's control.

That said, if dynamic hostname resolution is requested (to allow for multiple hostnames), the only secure way to do so is via properly configure proxy that sets the headers.

**Comment by sschu** [comment]
You can already override the host per-realm in the realm settings. I would just propose to allow a list of values there and Keycloak then uses the value coming from the host header if it is in this list.

**Comment by vmuzikar** [comment]
I see, basically a whitelist of allowed hostnames. That's an interesting idea...

**Comment by thomasdarimont** [comment]
The hostname configuration often takes a while to get right, and it is excellent to see some effort for simplification here.

One additional point to the things already mentioned:
Could we also get an `--hostname-backend`  option (as an addition to --hostname and --hostname-admin)?

Use-case:
The customer uses cloud-flare with bot detection, rate-limiting, etc., as a "frontend" for all user-facing URLs with "frontend" hostnames. However, backend applications of the customer should be able to access Keycloak with a different hostname (advertised in .well-known/openid-configuration), e.g., via Cloudflare tunnel. Of course, the customer could use the `openidConfigOverride` or implement a custom `OIDCWellKnownProvider` or a custom `HostnameProvider`, but this would not be necessary if we could also define a "fixed" backend hostname for the whole server or a realm.

**Comment by vmuzikar** [comment]
@thomasdarimont For backend, we have `--hostname-backchannel-dynamic`. In your stated use case, you'd set it to `true` which is meant for the cases when some apps are accessing Keycloak via internal addresses.

**Comment by ptrovatelli** [comment]
Hello,
Would it make sense to include in this work the possibility of configuring the issuer hostname? Currently the issuer,  advertised in the well-known url, is built by keycloak based on the **frontend hostname**. In my use case, it would make more sense for it to be buit based on the **backend hostname**, or even to make it fully configurable. 
See this discussion: https://groups.google.com/g/keycloak-user/c/xot6XrGCM8k
How about a `--hostname-issuer-url` parameter including protocol and port? The issuer would be `<hostname-issuer-url>/auth/realms/<realm>`

Thanks

**Comment by Tristan971** [comment]
Just in case this goes forward, quoting from that google groups link:

> end-users don't need to access the well-known url do they? 

They absolutely do; the well-known url spec is as useful for SPA/apps type frontends as it is for internal systems like Vault (quoting the disc. there again).

Now I do get your pain with Vault and other internal systems needing to go out of your infra, and onto the public internet before getting back in just to have it match. Not too sure about that being KC's responsibility though.
Here Vault should really be the ones to support overrides imo...

**Comment by tspearconquest** [comment]
Hello, if I'm having struggles on the current implementation and have hostname-debug output, should I make a new issue or post the output here, or on #14666 to get assistance troubleshooting? I came from #15018.

**Comment by vmuzikar** [comment]
@ptrovatelli I replied at the thread you sent. I believe your client is not implemented correctly. The issuer is a unique public identifier rather than just a URL and should be validated in the ID Token against whatever is present in .well-known endpoint.

@tspearconquest Please use https://github.com/keycloak/keycloak/issues/14666 as it's a single place for reporting issues  with v1 implementation.

**Comment by megian** [comment]
Why not arbitrary and multiple front-end paths per realm? From a security perspective it would be helpful to expose the same realm on different paths (internal, external) using different certificates. Is this something needs to be defined at startup or could this be handled according a configuration from the database?

Use cases:
* Exposing the same realm on https://keycloak.company.com and https://id.company.com
* Exposing the same realm on / and /auth
* Exposing a new realm on / while keeping existing ones on /auth
* Exposing a realm external on https://id.company.com and internal on https://keycloak.cluster.local (different TLS certificates)

**Comment by sunsided** [comment]
The recently closed #25230 ("Loading the Admin UI" forever) occurs for me on 23.0.6 when deployed using the Operator in Kubernetes, behind a port-forward. I use nip.io to fake a domain and provided ports `8080` and `8443`:

```yaml
apiVersion: k8s.keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: example-kc
  annotations:
    argocd.argoproj.io/sync-wave: "10"
spec:
  instances: 1
  image: quay.io/keycloak/keycloak:23.0.6-0
  db:
    vendor: postgres
    host: hippo-primary.postgres-operator.svc.cluster.local
    schema: keycloak
    usernameSecret:
      name: keycloak-db-secret
      key: user
    passwordSecret:
      name: keycloak-db-secret
      key: password
  http:
    httpEnabled: true
    httpPort: 8080
    httpsPort: 8443
    tlsSecret: keycloak-tls
  ingress:
    className: istio
  hostname:
    # can't set: url, port
    hostname: keycloak-127-1-1-7.nip.io
    adminUrl: https://admin.keycloak-127-1-1-7.nip.io:8443
    strict: false
    strictBackchannel: false
```

When used as is, visiting

- `https://keycloak-127-1-1-7.nip.io:8443/` gives "Welcome to Keycloak"
- `https://admin.keycloak-127-1-1-7.nip.io:8443/admin/master/console/` loads with "Loading the Admin UI"
- `https://keycloak-127-1-1-7.nip.io/realms/master/protocol/openid-connect/3p-cookies/step1.html` is attempted to be loaded (note the missing port)

If I then manually visit 

- `https://keycloak-127-1-1-7.nip.io:8443/realms/master/protocol/openid-connect/3p-cookies/step1.html` directly forwards to 
- `https://keycloak-127-1-1-7.nip.io:8443/realms/master/protocol/openid-connect/3p-cookies/step2.html`, but the page is empty.

According to [operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/HostnameSpec.java](https://github.com/keycloak/keycloak/blob/df38081fe89b7cde3896f9707c638fc2c36098c0/operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/HostnameSpec.java) I cannot specify the `hostname: url` nor `hostname: port` that is documented in https://www.keycloak.org/server/all-config#category-hostname.

**Comment by crstian19** [comment]
I have experienced the error ("Loading the Admin UI" forever) since 22.0.5, using an ingress and the variable: KC_PROXY= edge solved it.

Now in version 23.0.6, I want to not need an ingress to be able to use the admin console but I am not able to bypass this error,, I am using these variables and port forwarding to my port 8085:

```
          - name: KC_PROXY
            value: edge
          - name: KC_HOSTNAME_ADMIN_URL
            value: http://localhost:8085
          - name: KC_HOSTNAME_STRICT
            value: "false"
```

**Comment by heruan** [comment]
> Hello, Would it make sense to include in this work the possibility of configuring the issuer hostname? Currently the issuer, advertised in the well-known url, is built by keycloak based on the **frontend hostname**. In my use case, it would make more sense for it to be buit based on the **backend hostname**, or even to make it fully configurable. See this discussion: https://groups.google.com/g/keycloak-user/c/xot6XrGCM8k How about a `--hostname-issuer-url` parameter including protocol and port? The issuer would be `<hostname-issuer-url>/auth/realms/<realm>`

I'm facing a similar scenario where the frontend issuer hostname in the .well-known configuration can't be validated since the app requesting authentication fetches metadata from the backend hostname. This is the architecture:

```
 ______________________________________________________________________________
|        _______                                              ____________     |
|       |  APP  |  <== http://backend-hostname:8180/... ==>  |  KEYCLOAK  |    |
|        ¯¯||¯¯¯                                              ¯¯¯¯¯||¯¯¯¯¯     |
|__________||______________________________________________________||__________|
           ||                                                      ||
 https://app-hostname/                                https://frontend-hostname/
```

When APP (Java Spring Boot) fetches configuration from the .well-known endpoint, it uses the backend-hostname but metadata contains the `issuer` with the frontend-hostname and validation fails [as per spec](https://datatracker.ietf.org/doc/html/rfc8414#section-3.3):

>The "issuer" value returned MUST be identical to the authorization
server's issuer identifier value into which the well-known URI string
was inserted to create the URL used to retrieve the metadata. If
these values are not identical, the data contained in the response
MUST NOT be used.

Can the `issuer` value returned by the .well-known configuration be configured to use the hostname used for the request, so that validation can happen when fetching from the backend-hostname?

**Comment by vmuzikar** [comment]
Closing this epic as all issues in it were resolved.

We encourage everyone to give Hostname v2 a try!

**Comment by heruan** [comment]
Awesome, thanks @vmuzikar! I guess [my comment above](https://github.com/keycloak/keycloak/issues/24252#issuecomment-2018949211) was a bit late to catch that train. Any thoughts?

**Comment by vmuzikar** [comment]
@heruan This is a bit out of scope of Hostname SPI as its implementations just define how various types of URLs are resolved, not where they are used (i.e. if frontend URL type is used for issuer).

> Can the `issuer` value returned by the .well-known configuration be configured to use the hostname used for the request, so that validation can happen when fetching from the backend-hostname?

@keycloak/core can correct me (as this is their area – OIDC) but the issuer should be as static as possible to uniquely identify the auth server (OP in terms of OIDC). The same issuer (in our case based on the frontend URL) should be of course used both in the .well-known endpoint and in the tokens. If that was not the case in your use case, I believe it'd be a bug.

**Comment by ptrovatelli** [comment]
@heruan I saw the same issue with spring security than with vault yehp. See this issue: https://github.com/spring-projects/spring-security/issues/14633

maybe we should open a dedicated issue on keycloak side? it is a global problem, for all the clients that will be implementing sctrictly the oauth specification and when we have a front-end and a backend url on keycloak side. would be nice to have a way to bypass it on keycloak side. 
As far as as I know, the client doesn't usually validate that the issuer matches the well-known URL (because the client doesn't even need to fetch the well-known url) so using the backend URL instead of the frontend URL for the issuer should fix it
